File: implementations/react-native/design-patterns-rn/src/patterns/structural/flyweight/flyweight-factory.ts
import { BadgeFlyweight } from './concrete-flyweight';
import type { FlyweightBadge } from './flyweight-interface';

export class BadgeFactory {
  private cache = new Map<string, FlyweightBadge>();

  get(key: string): FlyweightBadge {
    if (!this.cache.has(key)) {
      const fw = this.create(key);
      this.cache.set(key, fw);
    }
    return this.cache.get(key)!;
  }

  size() {
    return this.cache.size;
  }

  private create(key: string): FlyweightBadge {
    switch (key) {
      case 'pill':
        return new BadgeFlyweight({
          borderRadius: 999,
          paddingHorizontal: 8,
          paddingVertical: 2,
          background: '#111',
          fontSize: 12,
          fontWeight: '600',
        });
      case 'rounded-outline':
        return new BadgeFlyweight({
          borderRadius: 8,
          paddingHorizontal: 8,
          paddingVertical: 2,
          borderWidth: 1,
          borderColor: '#333',
          background: '#fff',
          fontSize: 12,
          fontWeight: '600',
        });
      case 'chip':
      default:
        return new BadgeFlyweight({
          borderRadius: 14,
          paddingHorizontal: 10,
          paddingVertical: 0,
          background: '#f5f5f5',
          fontSize: 12,
          fontWeight: '500',
        });
    }
  }
}

File: implementations/flutter/design_patterns_flutter/lib/patterns/structural/flyweight/flyweight_factory.dart
import 'package:flutter/material.dart';
import 'flyweight_interface.dart';
import 'concrete_flyweight.dart';

class BadgeFactory {
  final Map<String, FlyweightBadge> _cache = {};

  FlyweightBadge get(String key) {
    if (!_cache.containsKey(key)) {
      _cache[key] = _create(key);
    }
    return _cache[key]!;
  }

  int size() => _cache.length;

  FlyweightBadge _create(String key) {
    switch (key) {
      case 'pill':
        return BadgeFlyweight(
          borderRadius: 999,
          padding: const EdgeInsets.symmetric(horizontal: 8, vertical: 2),
          border: null,
          background: const Color(0xFF111111),
          fontSize: 12,
          fontWeight: FontWeight.w600,
        );
      case 'rounded-outline':
        return BadgeFlyweight(
          borderRadius: 8,
          padding: const EdgeInsets.symmetric(horizontal: 8, vertical: 2),
          border: Border.all(color: const Color(0xFF333333)),
          background: Colors.white,
          fontSize: 12,
          fontWeight: FontWeight.w600,
        );
      case 'chip':
      default:
        return BadgeFlyweight(
          borderRadius: 14,
          padding: const EdgeInsets.symmetric(horizontal: 10, vertical: 0),
          border: null,
          background: const Color(0xFFF5F5F5),
          fontSize: 12,
          fontWeight: FontWeight.w500,
        );
    }
  }
}
